### ISSUE TITLE
link pdf files to notes

### ISSUE THREAD

**Issue by paynito** [jkitchin/org-ref#147]
I am using the melpa with the addition of the section:   ((Ok, now I see that's just a stub.))

    (defcustom org-ref-get-pdf-filename-function
       'org-ref-get-mendeley-filename
        "User-defined function to get a filename from a bibtex key.
        The function must take a key as an argument, and return the path
        to the corresponding filename.  The default is
        `org-ref-get-pdf-filename'.  An alternative value is
        `org-ref-get-mendeley-filename'."
        :type 'function
        :group 'org-ref)

here is the kind of .bib I'm working with

    @book{Bar-haim2015a,
    author = {Bar-haim, Eyal and Semyonov, Moshe},
    doi = {10.1007/978-90-481-8891-8},
    file = {:Users/me/Documents/pdfs/Bar-haim, Semyonov/2015/Bar-haimpaper.pdf:pdf},
    isbn = {978-90-481-8890-1},
    pages = {323--337},
    title = {{The International Handbook of the Demography of Race and Ethnicity}},
    year = {2015}
    }

ls -R tells me that Mendeley organizes my pdfs like this:

    ./Bar-Asher:
    2013
    2015

    ./Bar-Asher/2013:
    Bar-Asher - Astory - 2013.pdf

    ./Bar-Asher/2015:
    Bar-asher - World.pdf

    ./Bar-haim, Semyonov:
    2015

    ./Bar-haim, Semyonov/2015:
    Bar-haim, Semyonov -a.pdf
    Bar-haim, Semyonov - myPaper - 2015.pdf

----
So, I had wondered if setting the variables

     (setq helm-bibtex-library-path 

and
    
    org-ref-pdf-directory "~/Desktop/org-ref-example/bibtex-pdfs/")

to that directory would fix **my issue**, which is when I click on the  **org-ref-citation-link**
and choose option 
3 "open notes"
It connects my notes to ~/Desktop/org-ref-example/no pdf
but I want them to be with ~/Documents/pdfs/Twain, Mark/1864/HuckFinn.pdf
for example
## also##
should

   (setq helm-bibtex-notes-path "~/school/notes/helm-bibtex-notes") 

be set to the same path as

    (setq org-ref-bibliography-notes           

thank you again.

## tl;dr##
if I am making a **org-ref-citation-link** based on 
the cite key in
@book{Bar-haim2015a,

why isn't it helm seeing the file = 
field when I start taking notes?

**Comment by paynito** [comment]
This seems related.  `http://emacs.stackexchange.com/questions/19791/helm-bibtex-and-bibdesk-bibtex-intergration?rq=1`  so there is a competing "Bdsk-File-#"  not just "File:"  field.

**Comment by jkitchin** [comment]
The way to use this option is to put
`(setq org-ref-get-pdf-filename-function 'org-ref-get-mendeley-filename)`
In your init files before you load org-ref. 

Alternatively, you can write your own function that computes the pdf name from a bibtex key by some method.

To get links to non-standard pdf files in your notes, you may also have modify the variable at https://github.com/jkitchin/org-ref/blob/master/org-ref.el#L215 to use a function that finds your pdf and constructs a link to it.

**Comment by paynito** [comment]
Thanks John, I will try that now.

**Comment by paynito** [comment]
Leaving directory `/Users/me/.emacs.d/elpa/f-20151113.123'

Compiling file /Users/me/.emacs.d/elpa/f-20151113.123/f-pkg.el at Thu Mar  3 08:48:12 2016
Entering directory `/Users/me/.emacs.d/elpa/f-20151113.123/'

Compiling file /Users/me/.emacs.d/elpa/f-20151113.123/f.el at Thu Mar  3 08:48:12 2016

Compiling no file at Thu Mar  3 08:48:13 2016
Leaving directory `/Users/me/.emacs.d/elpa/key-chord-20160227.438'

Compiling file /Users/me/.emacs.d/elpa/key-chord-20160227.438/key-chord-pkg.el at Thu Mar  3 08:48:13 2016
Entering directory `/Users/me/.emacs.d/elpa/key-chord-20160227.438/'

Compiling file /Users/me/.emacs.d/elpa/key-chord-20160227.438/key-chord.el at Thu Mar  3 08:48:13 2016

In key-chord-input-method:
key-chord.el:341:10:Warning: sit-for called with 3 arguments, but accepts only
    1-2

In end of data:
key-chord.el:374:1:Warning: the function `eldoc-pre-command-refresh-echo-area'
    is not known to be defined.

Compiling no file at Thu Mar  3 08:48:13 2016
Leaving directory `/Users/me/.emacs.d/elpa/parsebib-20151006.232'

Compiling file /Users/me/.emacs.d/elpa/parsebib-20151006.232/parsebib-pkg.el at Thu Mar  3 08:48:13 2016
Entering directory `/Users/me/.emacs.d/elpa/parsebib-20151006.232/'

Compiling file /Users/me/.emacs.d/elpa/parsebib-20151006.232/parsebib.el at Thu Mar  3 08:48:13 2016

Compiling no file at Thu Mar  3 08:48:14 2016
Leaving directory `/Users/me/.emacs.d/elpa/helm-bibtex-20160225.1049'

Compiling file /Users/me/.emacs.d/elpa/helm-bibtex-20160225.1049/helm-bibtex-pkg.el at Thu Mar  3 08:48:14 2016
Entering directory `/Users/me/.emacs.d/elpa/helm-bibtex-20160225.1049/'

Compiling file /Users/me/.emacs.d/elpa/helm-bibtex-20160225.1049/helm-bibtex.el at Thu Mar  3 08:48:14 2016

In end of data:
helm-bibtex.el:999:1:Warning: the following functions are not known to be defined: show-all,
    org-narrow-to-subtree, org-cycle-hide-drawers

Compiling no file at Thu Mar  3 08:48:17 2016
Leaving directory `/Users/me/.emacs.d/elpa/org-ref-20160302.1411'

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/doi-utils.el at Thu Mar  3 08:48:17 2016
Entering directory `/Users/me/.emacs.d/elpa/org-ref-20160302.1411/'

In doi-utils-get-bibtex-entry-pdf:
doi-utils.el:474:33:Warning: reference to free variable
    `org-ref-pdf-directory'

In doi-utils-add-bibtex-entry-from-doi:
doi-utils.el:720:19:Warning: reference to free variable
    `org-ref-default-bibliography'

In doi-utils-open-bibtex:
doi-utils.el:941:16:Warning: reference to free variable
    `reftex-default-bibliography'

In doi-utils-add-entry-from-crossref-query:
doi-utils.el:1183:26:Warning: reference to free variable
    `org-ref-default-bibliography'

In end of data:
doi-utils.el:1227:1:Warning: the function `doi-utils-concat-prepare' might not
    be defined at runtime.
doi-utils.el:1227:1:Warning: the following functions are not known to be defined: reftex-get-bib-field,
    org-ref-bib-citation

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/nist-webbook.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-arxiv.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-bibtex.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-glossary.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-helm-bibtex.el at Thu Mar  3 08:48:18 2016
org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `fields'
org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `fields'
org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `fields'
org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `fields'
org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `entry'

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-helm-cite.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-helm.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-isbn.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-ivy-bibtex.el at Thu Mar  3 08:48:18 2016
org-ref-ivy-bibtex.el:26:1:Error: Cannot open load file: no such file or directory, swiper

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-latex.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-pdf.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-pubmed.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-reftex.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-sci-id.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-scifinder.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-scopus.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-url-utils.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-utils.el at Thu Mar  3 08:48:18 2016

In org-ref-debug:
org-ref-utils.el:138:17:Warning: reference to free variable
    `pdftotext-executable'

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-worldcat.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-wos.el at Thu Mar  3 08:48:18 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref.el at Thu Mar  3 08:48:18 2016

In org-ref-insert-key-at-point:
org-ref.el:2886:14:Warning: reference to free variable
    `org-ref-cite-completion-function'

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/x2bib.el at Thu Mar  3 08:48:18 2016
Leaving directory `/Users/me/.emacs.d/elpa/org-ref-20160302.1411/'

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/author-year.el at Thu Mar  3 08:48:18 2016
Entering directory `/Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/'

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/org-ref-citeproc.el at Thu Mar  3 08:48:18 2016

In orcp-citeproc:
org-ref-citeproc.el:869:17:Warning: looking-at called with 2 arguments, but
    accepts only 1

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/unsrt-footnote.el at Thu Mar  3 08:48:19 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/unsrt-paren.el at Thu Mar  3 08:48:19 2016

Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/unsrt.el at Thu Mar  3 08:48:19 2016

Compiling no file at Thu Mar  3 08:48:19 2016

**Comment by paynito** [comment]
ok, M-x
org-autocite-complete-link   -> broken  doesn't insert anything
org-ref-insert-cite-with-complete -> working, inserts something
org-ref-helm-insert-cite-link  -> working, inserts something

but
now when i click the green [cite:Twain1864](www.bing.com)

it flashes wrong type argument listp, ":Users/me/pdfs/Twain/1864/huck.pdf"

so, I've stuck the 

    (setq org-ref-get-pdf-filename-function 'org-ref-get-mendeley-filename)

in the wrong place it seems.  Maybe I should look at where you had it in your blob.el and put it there in my melpa.el

**Comment by jkitchin** [comment]
org-autocite-complete-link is for use with C-c C-l to insert links, it is not used with M-x.

I guess you need a list of paths in your bibtex entry, and you only have a string, and the mendeley function is complaining that it expected a list, and got a string.

**Comment by jkitchin** [comment]
I tried out the mendeley file name function today and was able to reproduce the error you reported. I have pushed a modification that fixes that error. It works for a test file I generated from Mendeley. I don't know if it works for all Mendeley files though, e.g. if there are more than one pdf stored in an entry. I don't have any examples like that. I think it will, but am not sure. Let me know if the new fix helps!

Also, I made all the org-*-complete-link functions not interactive so they don't get confused. They should be used with the org-mode insert link functions or C-c C-l.

**Comment by paynito** [comment]
Thank you, will the same melpa.el I use download the new changes?  I guess that's how it works. 
I have also been getting an error about some intermediate compile file not being in the org-ref-example/sandbox folder. Thanks for the fix. I was ready to just switch all my files to citekey.pdf naming and a flat folder to make org work. I really hate that when I fix my bib data, the author, Title Case, the book and chapter and publication city and everything that you seem to have to hand correct on every pdf that comes without a DOI, that Mendeley doesn't save my corrected meta-data back into the pdf. I know that some are protected, but . . . Dr. Kitchin you have created the best-in-class software for what we do right now. Congratulations. I'm interested in another screen inside emacs that lets me view one bib entry at a time, like the view in BibDesk. So I can tab between fields and make corrections. When I visit the whole .bib from org-ref I don't remember the field names to add them. I mean that often I have a book chapter that is mislabelled in my .bib as an @article entry.  Then I have to change it to @incollection and then add the :BookTitle Field.  Anyway, I guess helm can simulate a drop-down menu that let's you click on @article and choose from the list of other types or just free type your choice.  Thanks again.
 I really liked the one lecture you did with the 12k views. It was inspiring. I will be doing a stats project w/ some census data and hope to use org-mode now w/ either sci-py or R.

**Comment by jkitchin** [comment]
You can ignore all these warnings.

paynito writes:

> Leaving directory `/Users/me/.emacs.d/elpa/f-20151113.123'
> 
> Compiling file /Users/me/.emacs.d/elpa/f-20151113.123/f-pkg.el at Thu Mar  3 08:48:12 2016
> Entering directory `/Users/me/.emacs.d/elpa/f-20151113.123/'
> 
> Compiling file /Users/me/.emacs.d/elpa/f-20151113.123/f.el at Thu Mar  3 08:48:12 2016
> 
> Compiling no file at Thu Mar  3 08:48:13 2016
> Leaving directory `/Users/me/.emacs.d/elpa/key-chord-20160227.438'
> 
> Compiling file /Users/me/.emacs.d/elpa/key-chord-20160227.438/key-chord-pkg.el at Thu Mar  3 08:48:13 2016
> Entering directory `/Users/me/.emacs.d/elpa/key-chord-20160227.438/'
> 
> Compiling file /Users/me/.emacs.d/elpa/key-chord-20160227.438/key-chord.el at Thu Mar  3 08:48:13 2016
>
> In key-chord-input-method:
> key-chord.el:341:10:Warning: sit-for called with 3 arguments, but accepts only
>     1-2
>
> In end of data:
> key-chord.el:374:1:Warning: the function `eldoc-pre-command-refresh-echo-area'
>     is not known to be defined.
> 
> Compiling no file at Thu Mar  3 08:48:13 2016
> Leaving directory `/Users/me/.emacs.d/elpa/parsebib-20151006.232'
> 
> Compiling file /Users/me/.emacs.d/elpa/parsebib-20151006.232/parsebib-pkg.el at Thu Mar  3 08:48:13 2016
> Entering directory `/Users/me/.emacs.d/elpa/parsebib-20151006.232/'
> 
> Compiling file /Users/me/.emacs.d/elpa/parsebib-20151006.232/parsebib.el at Thu Mar  3 08:48:13 2016
> 
> Compiling no file at Thu Mar  3 08:48:14 2016
> Leaving directory `/Users/me/.emacs.d/elpa/helm-bibtex-20160225.1049'
> 
> Compiling file /Users/me/.emacs.d/elpa/helm-bibtex-20160225.1049/helm-bibtex-pkg.el at Thu Mar  3 08:48:14 2016
> Entering directory `/Users/me/.emacs.d/elpa/helm-bibtex-20160225.1049/'
> 
> Compiling file /Users/me/.emacs.d/elpa/helm-bibtex-20160225.1049/helm-bibtex.el at Thu Mar  3 08:48:14 2016
>
> In end of data:
> helm-bibtex.el:999:1:Warning: the following functions are not known to be defined: show-all,
>     org-narrow-to-subtree, org-cycle-hide-drawers
> 
> Compiling no file at Thu Mar  3 08:48:17 2016
> Leaving directory `/Users/me/.emacs.d/elpa/org-ref-20160302.1411'
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/doi-utils.el at Thu Mar  3 08:48:17 2016
> Entering directory `/Users/me/.emacs.d/elpa/org-ref-20160302.1411/'
>
> In doi-utils-get-bibtex-entry-pdf:
> doi-utils.el:474:33:Warning: reference to free variable
>     `org-ref-pdf-directory'
>
> In doi-utils-add-bibtex-entry-from-doi:
> doi-utils.el:720:19:Warning: reference to free variable
>     `org-ref-default-bibliography'
>
> In doi-utils-open-bibtex:
> doi-utils.el:941:16:Warning: reference to free variable
>     `reftex-default-bibliography'
>
> In doi-utils-add-entry-from-crossref-query:
> doi-utils.el:1183:26:Warning: reference to free variable
>     `org-ref-default-bibliography'
>
> In end of data:
> doi-utils.el:1227:1:Warning: the function `doi-utils-concat-prepare' might not
>     be defined at runtime.
> doi-utils.el:1227:1:Warning: the following functions are not known to be defined: reftex-get-bib-field,
>     org-ref-bib-citation
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/nist-webbook.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-arxiv.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-bibtex.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-glossary.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-helm-bibtex.el at Thu Mar  3 08:48:18 2016
> org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `fields'
> org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `fields'
> org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `fields'
> org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `fields'
> org-ref-helm-bibtex.el:96:1:Warning: Unused lexical variable `entry'
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-helm-cite.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-helm.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-isbn.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-ivy-bibtex.el at Thu Mar  3 08:48:18 2016
> org-ref-ivy-bibtex.el:26:1:Error: Cannot open load file: no such file or directory, swiper
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-latex.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-pdf.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-pubmed.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-reftex.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-sci-id.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-scifinder.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-scopus.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-url-utils.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-utils.el at Thu Mar  3 08:48:18 2016
>
> In org-ref-debug:
> org-ref-utils.el:138:17:Warning: reference to free variable
>     `pdftotext-executable'
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-worldcat.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref-wos.el at Thu Mar  3 08:48:18 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/org-ref.el at Thu Mar  3 08:48:18 2016
>
> In org-ref-insert-key-at-point:
> org-ref.el:2886:14:Warning: reference to free variable
>     `org-ref-cite-completion-function'
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/x2bib.el at Thu Mar  3 08:48:18 2016
> Leaving directory `/Users/me/.emacs.d/elpa/org-ref-20160302.1411/'
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/author-year.el at Thu Mar  3 08:48:18 2016
> Entering directory `/Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/'
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/org-ref-citeproc.el at Thu Mar  3 08:48:18 2016
>
> In orcp-citeproc:
> org-ref-citeproc.el:869:17:Warning: looking-at called with 2 arguments, but
>     accepts only 1
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/unsrt-footnote.el at Thu Mar  3 08:48:19 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/unsrt-paren.el at Thu Mar  3 08:48:19 2016
> 
> Compiling file /Users/me/.emacs.d/elpa/org-ref-20160302.1411/citeproc/unsrt.el at Thu Mar  3 08:48:19 2016
> 
> Compiling no file at Thu Mar  3 08:48:19 2016
>
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/jkitchin/org-ref/issues/147#issuecomment-191613876


--
Professor John Kitchin
Doherty Hall A207F
Department of Chemical Engineering
Carnegie Mellon University
Pittsburgh, PA 15213
412-268-7803
@johnkitchin
http://kitchingroup.cheme.cmu.edu
